File: problems/0036.valid-sudoku/valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
const isValidSudoku = function (board) {
  let [row] = board;
  let column = row.map((_, column) => board.map((row) => row[column]));
  let sub_boxes = Array(9).fill([]);
  let valid = false;

  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board.length; j++) {
      let boxIndex = Math.floor(i / 3) * 3 + Math.floor(j / 3);
      sub_boxes[boxIndex] = [...sub_boxes[boxIndex], board[i][j]];
    }
  }

  for (let i = 0; i < board.length; i++) {
    valid =
      !board[i].some(
        (r, index) => r !== "." && board[i].indexOf(r) !== index
      ) &&
      !column[i].some(
        (col, index) => col !== "." && column[i].indexOf(col) !== index
      ) &&
      !sub_boxes[i].some(
        (sub, index) => sub !== "." && sub_boxes[i].indexOf(sub) !== index
      );
    if (!valid) return false;
  }

  return valid;
};

File: problems/0036.valid-sudoku/valid_sudoku.py
from ast import List


class Solution:
    def is_valid_sudoku(self, board: List[List[str]]) -> bool:
        """
        Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:

        Each row must contain the digits 1-9 without repetition.
        Each column must contain the digits 1-9 without repetition.
        Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.
        Note:

        A Sudoku board (partially filled) could be valid but is not necessarily solvable.
        Only the filled cells need to be validated according to the mentioned rules.

        Args:
        board (List[List[str]]): A 9x9 grid representing the Sudoku board.

        Returns:
        bool: True if the board is valid, False otherwise.
        """
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for row in range(9):
            for col in range(9):
                num = board[row][col]
                
                if num == ".":
                    continue
                
                box_index = (row // 3) * 3 + (col // 3)

                if num in rows[row]:
                    return False
                if num in cols[col]:
                    return False
                if num in boxes[box_index]:
                    return False

                rows[row].add(num)
                cols[col].add(num)
                boxes[box_index].add(num)

        return True